backends/cuda/ceed-cuda-restriction.c

```diff
@@ -0,0 +1,252 @@
+  CeedElemRestriction_Cuda *impl;
+  ierr = CeedCalloc(1, &impl); CeedChk(ierr);
+  CeedInt nelem, elemsize;
+  ierr = CeedElemRestrictionGetNumElements(r, &nelem); CeedChk(ierr);
+  ierr = CeedElemRestrictionGetElementSize(r, &elemsize); CeedChk(ierr);
+  CeedInt size = nelem * elemsize;
+
+  ierr = CeedCalloc(1, &impl); CeedChk(ierr);
+  impl->h_ind           = NULL;
+  impl->h_ind_allocated = NULL;
+  impl->d_ind           = NULL;
+  impl->d_ind_allocated = NULL;
+
+  if (mtype == CEED_MEM_HOST) {
+    switch (cmode) {
+    case CEED_OWN_POINTER:
+      impl->h_ind_allocated = (CeedInt *)indices;
+    case CEED_USE_POINTER:
+      impl->h_ind = (CeedInt *)indices;
+    case CEED_COPY_VALUES:
+      if (indices != NULL) {
+        ierr = cudaMalloc( (void**)&impl->d_ind, size * sizeof(CeedInt));
+        CeedChk_Cu(ceed, ierr);
+        impl->d_ind_allocated = impl->d_ind;//We own the device memory
+        ierr = cudaMemcpy(impl->d_ind, indices, size * sizeof(CeedInt), cudaMemcpyHostToDevice);
+        CeedChk_Cu(ceed, ierr);
+      }
+    }
+  } else if (mtype == CEED_MEM_DEVICE) {
+    switch (cmode) {
+    case CEED_COPY_VALUES:
+      if (indices != NULL) {
+        ierr = cudaMalloc( (void**)&impl->d_ind, size * sizeof(CeedInt));
+        CeedChk_Cu(ceed, ierr);
+        impl->d_ind_allocated = impl->d_ind;//We own the device memory
+        ierr = cudaMemcpy(impl->d_ind, indices, size * sizeof(CeedInt), cudaMemcpyDeviceToDevice);
+        CeedChk_Cu(ceed, ierr);
+      }
+      break;
+    case CEED_OWN_POINTER:
+      impl->d_ind = (CeedInt *)indices;
+      impl->d_ind_allocated = impl->d_ind;
+      break;
+    case CEED_USE_POINTER:
+      impl->d_ind = (CeedInt *)indices;
+    }
+  } else
+    return CeedError(ceed, 1, "Only MemType = HOST or DEVICE supported");
+
+  CeedInt ncomp, ndof;
+  ierr = CeedElemRestrictionGetNumComponents(r, &ncomp); CeedChk(ierr);
+  ierr = CeedElemRestrictionGetNumDoF(r, &ndof); CeedChk(ierr);
+  ierr = compile(ceed, restrictionkernels, &impl->module, 3,
+                 "RESTRICTION_ELEMSIZE", elemsize,
+                 "RESTRICTION_NCOMP", ncomp,
+                 "RESTRICTION_NDOF", ndof); CeedChk(ierr);
+  ierr = get_kernel(ceed, impl->module, "noTrNoTr", &impl->noTrNoTr); CeedChk(ierr);
+  ierr = get_kernel(ceed, impl->module, "noTrTr", &impl->noTrTr); CeedChk(ierr);
+  ierr = get_kernel(ceed, impl->module, "trNoTr", &impl->trNoTr); CeedChk(ierr);
+  ierr = get_kernel(ceed, impl->module, "trTr", &impl->trTr); CeedChk(ierr);
+
+  ierr = CeedElemRestrictionSetData(r, (void*)&impl); CeedChk(ierr);
+  ierr = CeedSetBackendFunction(ceed, "ElemRestriction", r, "Apply",
+                                CeedElemRestrictionApply_Cuda); CeedChk(ierr);
+  ierr = CeedSetBackendFunction(ceed, "ElemRestriction", r, "Destroy",
+                                CeedElemRestrictionDestroy_Cuda); CeedChk(ierr);
+  return 0;
+}
+
+int CeedElemRestrictionCreateBlocked_Cuda(const CeedMemType mtype,
```

Comment:
This is just a question to improve my knowledge: We use blocked restrictions to help us interlace dofs and get faster BasisApply. Would this help CUDA/GPUs at all?

Reply:
This help if you want to hide the Restriction/Prolongation on the CPU. Then you treat elements at E-vector level on GPU, and do the Restriction/Prolongation on CPU. If you do everything on GPU it's unclear to me that there is any benefit.